MusicNotationKit/MusicNotationKit/Tuplet.swift

```diff
@@ -117,72 +132,338 @@ public struct Tuplet: NoteCollection {
+        } else {
+            self = toModify
+        }
+    }
+
+    private mutating func insert(_ noteCollections: [NoteCollection], at flatIndex: [Int]) throws {
+        if flatIndex.count != 1 {
+            // recurse to get to actual tuplet
+            let sliced = Array(flatIndex.dropFirst())
+            guard var tuplet = notes[flatIndex[0]] as? Tuplet else {
+                assertionFailure("all indexes before the last should be tuplets. Must be an error in flatIndexes")
+                throw TupletError.internalError
+            }
+            try tuplet.insert(noteCollections, at: sliced)
+            notes[flatIndex[0]] = tuplet
+        } else {
+            notes.insert(contentsOf: noteCollections, at: flatIndex[0])
+        }
+    }
+
+    private mutating func removeNotesRecursive(at flatIndexes: [[Int]]) throws {
+        guard let first = flatIndexes.first else {
+            return
+        }
+        // Need to get to the level where the final tuplet is
+        if first.count > 2 {
+            // recurse to get to actual tuplet level
+            let sliced = Array(flatIndexes.map { Array($0.dropFirst()) })
+            guard var tuplet = notes[first[0]] as? Tuplet else {
+                assertionFailure("all indexes before the last should be tuplets. Must be an error in flatIndexes")
+                throw TupletError.internalError
+            }
+            try tuplet.removeNotesRecursive(at: sliced)
+            // Set modified tuplet in enclosing tuplet's notes
+            notes[first[0]] = tuplet
+        } else if first.count == 2 {
+            // This is the same as the method that calls this. Should be able to be extracted into helper function
+            let tupletIndexes = Set(flatIndexes.map { $0[0] })
+            let tupletGroups = tupletIndexes.map { tupletIndex in
+                flatIndexes.filter { group in
+                    return group[0] == tupletIndex
+                }
+            }
+            for tupletGroup in tupletGroups {
+                let removeRange = tupletGroup[0][1]...tupletGroup[tupletGroup.count - 1][1]
+                guard var tuplet = notes[tupletGroup[0][0]] as? Tuplet else {
+                    assertionFailure("all indexes before the last should be tuplets. Must be an error in flatIndexes")
+                    throw TupletError.internalError
+                }
+                tuplet.notes.removeSubrange(removeRange)
+                notes[tupletGroup[0][0]] = tuplet
+            }
+        } else {
+            // They are not compound tuplets
+            // Just remove from the notes
+            let removeRange = flatIndexes[0][0]...flatIndexes[flatIndexes.count - 1][0]
+            notes.removeSubrange(removeRange)
+        }
     }
 
-    internal mutating func replaceNotes(in range: Range<Int>, with noteCollection: NoteCollection) throws {
-        throw NSError()
+    private mutating func removeNotes(at flatIndexes: [[Int]]) throws {
```

Comment:
If the `flatIndexes` are ordered, then we could traverse in reverse order  and create a `recursiveRemoveNote`  that removes single notes. If the result tuplet is empty, then we can remove the tuplet container. We can try to do this later if is too much trouble to do now.